packages/playwright-test/src/runner/uiMode.ts

```diff
@@ -14,112 +14,144 @@
+      const { method, params }: { method: string, params: any } = data;
+      if (method === 'list')
+        await this._listTests();
+      if (method === 'run')
+        await this._runTests(params.testIds);
+      if (method === 'stop')
+        this._stopTests();
+      if (method === 'watch')
+        this._watchFile(params.fileName);
+      if (method === 'exit')
+        exitPromise.resolve();
+    });
+    await exitPromise;
+  }
+
+  private _dispatchEvent(message: any) {
+    // eslint-disable-next-line no-console
+    this._page.mainFrame().evaluateExpression(dispatchFuncSource, true, message).catch(e => console.log(e));
   }
 
-  // Show trace viewer.
-  const page = await showTraceViewer([], 'chromium', { watchMode: true });
-  await page.mainFrame()._waitForFunctionExpression(serverSideCallMetadata(), '!!window.dispatch', false, undefined, { timeout: 0 });
-  {
-    // List
-    const controller = new Controller(config, page);
-    const listReporter = new TeleReporterEmitter(message => controller!.send(message));
+  private async _listTests() {
+    const listReporter = new TeleReporterEmitter(e => this._dispatchEvent(e));
     const reporter = new Multiplexer([listReporter]);
-    const taskRunner = createTaskRunnerForList(config, reporter);
-    const context: TaskRunnerState = {
-      config,
-      reporter,
-      phases: [],
-    };
-    reporter.onConfigure(config);
-    const { status, cleanup: globalCleanup } = await taskRunner.runDeferCleanup(context, 0);
-    if (status !== 'passed')
-      return await globalCleanup();
+    const taskRunner = createTaskRunnerForList(this._config, reporter);
+    const context: TaskRunnerState = { config: this._config, reporter, phases: [] };
+    reporter.onConfigure(this._config);
     await taskRunner.run(context, 0);
   }
 
-  await new Promise(() => {});
-  // TODO: implement watch queue with the sigint watcher and global teardown.
-  return 'passed';
-}
+  private async _runTests(testIds: string[]) {
+    await this._stopTests();
 
+    const testIdSet = testIds ? new Set<string>(testIds) : null;
+    this._config._internal.testIdMatcher = id => !testIdSet || testIdSet.has(id);
 
-class Controller {
-  private _page: Page;
-  private _queue = Promise.resolve();
-  private _runReporter: TeleReporterEmitter;
+    const runReporter = new TeleReporterEmitter(e => this._dispatchEvent(e));
+    const reporter = new Multiplexer([new ListReporter(), runReporter]);
+    const taskRunner = createTaskRunnerForWatch(this._config, reporter);
+    const context: TaskRunnerState = { config: this._config, reporter, phases: [] };
+    clearCompilationCache();
+    reporter.onConfigure(this._config);
+    const stop = new ManualPromise();
+    const run = taskRunner.run(context, 0, stop).then(async status => {
+      await reporter.onExit({ status });
+      return status;
+    });
+    this._testRun = { run, stop };
+    await run;
+    this._testRun = undefined;
```

Comment:
I'd move this line to 116.5, so that when `stop` awaits `testRun.run` we would be sure there is no run anymore.